woocommerce/woocommerce-admin
# Use Route based code splitting to optimize bundle size

Implemented [code splitting](https://reactjs.org/docs/code-splitting.html) on each route so only the code needed to display the current route is loaded on the page. Greatly benefits pages which only show the header. Uncompressed JS+CSS size reduced from 926 kb to 627 kb.

Additionally, minification has been added to the production build - resulting in `*.min.js` files.

Further optimizations could be done if we split up `@woocommerce/components` but probably that's for a seperate PR. 

Also creates the command `npm run analyze` to see a breakdown of each bundle and what is inside of it.

### Screenshots
Header Only page on master:
![header](https://user-images.githubusercontent.com/1534605/78860366-46c9db00-79ef-11ea-8cbe-189bf1dfc869.png)
Header only page on optimize-bundles:
![Screenshot from 2020-04-08 23-54-41](https://user-images.githubusercontent.com/1534605/78862360-75968000-79f4-11ea-9980-862ce047d008.png)
Dashboard page on master:
![dashboard](https://user-images.githubusercontent.com/1534605/78860365-46314480-79ef-11ea-91fb-b22e89634fd1.png)
Dashboard page on optimize-bundles:
![dashboard optimized](https://user-images.githubusercontent.com/1534605/78860360-45001780-79ef-11ea-822d-d63736effa7f.png)

### Detailed test instructions:

- Ensure `SCRIPT_DEBUG` is `true` in your `wp-config.php`
- Verify non-minified JS is enqueued (check network tab and page source)
- Navigate to all the pages and ensure they still load (smoke test everything)
- Ensure `SCRIPT_DEBUG` is `false` in your `wp-config.php`
- Verify minified JS is enqueued (check network tab and page source)
- Navigate to all the pages and ensure they still load (smoke test everything)
- Verify that `*.js` and `*.min.js` files are generated in both `development` and `production` modes (check that prod `min` files are minified)

### Changelog Note:

Performance: Implemented code splitting and lazy loading.
